src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs

```diff
@@ -67,30 +67,40 @@ internal sealed class NodeStateTable<T> : IStateTable
 
         private NodeStateTable(ImmutableArray<ImmutableArray<(T, EntryState)>> states, bool isCompacted, Exception? exception)
         {
+            Debug.Assert(exception is object && !isCompacted
+                        || exception is null && isCompacted && states.All(s => s.All(e => e.Item2 == EntryState.Cached))
```

Comment:
>&& isCompacted && states.All(s => s.All(e => e.Item2 == EntryState.Cached)) [](http://example.com/codeflow?start=45&length=75)

Is this catching any cases that aren't caught by `|| exception is null` below?

Reply:
Doing these multi boolean asserts always makes my brain hurt, but its aiming to say that isCompacted implies all the entries are cached (or empty). 

Would
```csharp 
Debug.Assert(exception is object && !isCompacted
            || exception is null && (!isCompacted || isCompacted && states.All(s => s.All(e => e.Item2 == EntryState.Cached))));
``` 
be clearer?


---
In reply to: [632175081](https://github.com/dotnet/roslyn/pull/53383#discussion_r632175081) [](ancestors = 632175081)